text/0000-prelude-2021.md

```diff
@@ -0,0 +1,164 @@
+
+While a number of third-party hashmaps are in common use (most of which reexport or wrap the std `HashMap` with a different hasher), bringing the std `HashMap` into the prelude should not cause issues, since explicit imports will always override prelude imports.
+
+## `std::{{fs::File}, path::{Path, PathBuf}}`
+
+Are very commonly used, not least of which in script-like CLI tools.
+
+## `std::sync::{Arc, Mutex, RwLock}`
+
+Help write threaded code. A potential concern is that async code may want to use async variants of these, although even then the [sync Mutex](https://docs.rs/tokio/1.1.1/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) may provide better performance.
+
+Use of alternative implementations (in particular from parking_lot) is fairly common.
+
+## `std::borrow::Cow`
+
+Unlikely to clash with other types, yet this type can be of great help in bridging the gap between references and owned types, particular for `String`/`str` and `Vec`/`slice` types.
+
+## `std::cmp::{min, max}`
+
+It has come up in [discussions](https://internals.rust-lang.org/t/am-i-the-only-one-confused-by-a-min-b-and-a-max-b/13252) that the method forms of `min()` and `max()` attached to std numeric types can be confusing to understand. In order to level the playing field with the two-argument free functions of the same names, it would be helpful to import these into the prelude.
+
+## Other traits that have been suggested for inclusion
+
+Other traits that have been suggested as prelude candidates:
+
+- `std::ops::Not`: for chaining purposes, it is sometimes useful to have an trailing `.not()` call rather than the prefix `!` operator. Therefore, [it has been suggested](https://internals.rust-lang.org/t/pre-rfc-adding-bool-not-method/13935) that the `Not` trait which brings that method in scope could be added to the prelude. This RFC author feels this use case is better served by adding an inherent impl to `bool`, since that serves the majority of the same use case with less complexity.
+- `std::fmt::Display`: users of this trait generally don't need to import it because the `ToString` trait which relies on the `Display` implementation is already in the prelude. Implementers of the `Display` trait however need several other items from `std::fmt` to do so; therefore, just importing `Display` into the prelude does not help much. This RFC suggests adding `fmt` to the prelude to improve this use case.
+- `std::fmt::Debug`: similar to `Display`, although there's no `to_debug()`. However, usage will usually go through `dbg!()` or the formatting mechanism (as a `{:?}` format string).
+- `std::future::Future`: `Future`'s `poll()` method is usually not called directly, but most often used via `async`/`await`, therefore including `Future` in the prelude does not seem as useful.
+- `std::error::Error`: while this is commonly implemented, calling its methods directly does not seem to be common enough that adding this is as useful.
```

Comment:
Another drawback: this would be ambiguous with `std::io::Error`.

Reply:
yeah calling both types Error was a big flub :/

Reply:
@Lokathor it's a rather common design choice to reuse names in different modules. This means that you have to write `io::Error` instead of something like `IoError`, which I actually prefer. Another example is `fmt::Write` and `io::Write`. However, I don't like that the [`write!`](https://doc.rust-lang.org/std/macro.write.html) macro requires importing one of them, but that's another topic.

Reply:
A common misconception. Only std::io and std::fmt do that weird thing where you're supposed to use the module name as a prefix. Every other error in the standard library just has a sensible name that doesn't need a prefix.

Either way, it's a bad plan at this point to put something called `Error` directly in the prelude.

It's *possible* that the confusion wouldn't be too high because one is a trait and the others are structs, but better to avoid the possibility.

Reply:
> Why not add std::sync to the prelude like it was proposed for std::mem?

I think `sync` is probably as reasonable as the other proposed modules, but I honestly think you could make that argument for almost every module in the stdlib. I htink for that it's a compelling example of where having some sort of policy here would help.

Reply:
The prelude could include something like
```rust
pub type StdError = std::error::Error
```
(Exact name to be bikeshedded)

Reply:
...that's absolutely horrifying.

Reply:
Perhaps @yaahc or other members of the error working group have input there.

Reply:
Personally I actually like the idea of having `StdError` be available, as this is a pattern I've seen in a number of crates (and have also used it in some of my own), but I'd definitely be interested in feedback from the erorr WG on this.

Reply:
IMHO, may as well rename the current Error into IOError through a typedef and expose the std error as Error.

Reply:
> Perhaps @yaahc or other members of the error working group have input there.

I'll have to think about this a bit more but my first reaction is that I don't want to include `std::error::Error` in the prelude or add an alias like `StdError`. I could see myself bending on adding an alias, though I really don't want it to be called `StdError`, that just seems strange to me, since no other std types have `Std` in their identifier other than `Stdin` and `Stdout` and in those it has a different meaning.

I think the most important question we need to ask is in what situations people people find themselves needing to refer to `std::error::Error` enough times in the same file where they'd like it to be shortened for them. In my mind the most likely explanation is for files that heavily use `Box<dyn std::error::Error + ..>`. If this is the only reason then I'm against adding an alias because I'm already working on making this a non issue.

The new `Try` rfc makes it possible to separate `From<E: Error>` from the error types being propagated and instead make the `From` conversion part of the `Result` type being used. The lack of this is what prevents `Box<dyn Error>` from implementing the `Error` trait itself, so I'm working on a proposal to add a `Box<dyn Error>` equivalent which is essentially identical except that it lacks the `From` impl and _does_ implement `Error`, along with a set of lints to help people transition from using `Box<dyn Error>` directly to instead use the fixed newtype. If we manage to untangle the type erased error issues then we should hopefully only need to name the error trait when defining it, and to me it seems fine to just write out the fully qualified name in these cases, or use derives that do so for you.

Reply:
I agree on the trait not getting an alias.

But the types that would be confused for the trait could get aliases: IoError and FmtError.

Reply:
> But the types that would be confused for the trait could get aliases: IoError and FmtError.

I don't think this addresses the problem here, which is that `Error` is a commonly reused type name, users are still going to have all sorts of their own error types that conflict with the error trait. Also the difference between `IoError` and `io::Error` seem small enough to me to not justify an alias.

Reply:
Well, I suppose. But users should also stop calling things just "Error". I suppose that ship has sailed.

Reply:
I don’t think it’s useful to blame users for following the standard library’s pattern of `io::Error` and `fmt::Error`.

Reply:
True, but the standard library has since moved away from that pattern (AllocError, Utf8Error, etc).

Reply:
AllocError is not stable and there's an open point about renaming it. Utf8Error is not in an utf8 module, so it's not changing any conventions.

Reply:
Personally I find it annoying that I have to write `std::sync::atomic::{AtomicU8, AtomicU16}` instead of `use std::sync::atomic` and then `atomic::U8` and `atomic::U16` (it's even worse to write `atomic::AtomicU8` and `atomic::AtomicU16`). I think it is quite idiomatic in the current state of Rust that we always prefix external items with their module (or crate if it's top-level), similar to the `package.Item` convention in golang (in which you can't even avoid that).

Reply:
Yeah, I kind of wish atomics were just in `std::sync::AtomicU8` instead of `std::sync::atomic::AtomicU8`. I don't think `atomic::U8` would be an improvement though, honestly I feel that would be worse. (This isn't on-topic though. These discussions get long enough as is)

> I think it is quite idiomatic in the current state of Rust that we always prefix external items with their module 

I don't think this bears out in practice. Most of the code I see and interact with isn't written this way. Additionally, autoimport functionality in IDEs doesn't work this way.